File: task3.py
import copy
import math

MAX_ITER = 10000
EPS = 1e-3

def gauss(A_in, b_in):
    n = len(A_in)
    A = copy.deepcopy(A_in)
    b = b_in[:]
    x = [0.0] * n

    # Прямой ход
    for k in range(n):
        pivotRow = k
        for i in range(k+1, n):
            if abs(A[i][k]) > abs(A[pivotRow][k]):
                pivotRow = i
        A[k], A[pivotRow] = A[pivotRow], A[k]
        b[k], b[pivotRow] = b[pivotRow], b[k]

        for i in range(k+1, n):
            factor = A[i][k] / A[k][k]
            for j in range(k, n):
                A[i][j] -= factor * A[k][j]
            b[i] -= factor * b[k]

    # Обратный ход
    for i in range(n-1, -1, -1):
        s = b[i]
        for j in range(i+1, n):
            s -= A[i][j] * x[j]
        x[i] = s / A[i][i]

    return x

def zeidel(A, b):
    n = len(A)
    x = [0.0] * n
    prev = [0.0] * n

    for _ in range(MAX_ITER):
        for i in range(n):
            if abs(A[i][i]) < 1e-12:
                print("Итерационный процесс не сошелся: нулевой диагональный элемент!")
                return []
            s = b[i]
            for j in range(i):
                s -= A[i][j] * x[j]
            for j in range(i+1, n):
                s -= A[i][j] * prev[j]
            x[i] = s / A[i][i]
            if math.isnan(x[i]):
                print("Итерационный процесс не сошелся: NaN!")
                return []

        eps = max(abs(x[i] - prev[i]) for i in range(n))
        if eps < EPS:
            return x
        prev = x[:]

    print(f"Итерационный процесс не сошелся за {MAX_ITER} итераций!")
    return []

def main():
    M = 0.93
    N = 0.07
    P = -0.84
    A = [
        [M,   -0.04, 0.21, -18],
        [0.25, -1.23, N,   -0.09],
        [-0.21, N,    0.8, -0.13],
        [0.15, -1.31, 0.06,  P]
    ]
    b = [-1.24, P, 2.56, M]

    print("Решение методом Гаусса:")
    sol_g = gauss(A, b)
    for val in sol_g:
        print(f"{val:6}", end=" ")
    print()

    print("Решение методом Зейделя:")
    sol_z = zeidel(A, b)
    for val in sol_z:
        print(f"{val:6}", end=" ")
    print()

if __name__ == "__main__":
    main()

File: task3.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>

using namespace std;

typedef vector<vector<double>> Matrix;
typedef vector<double> Line;

const int MAX_ITER = 10000;
const double EPS = 1e-3;
const int PRINT_STEP = 50;

Line gauss(const Matrix& A_in, const Line& b_in) {
    int n = A_in.size();
    Matrix A = A_in;
    Line b = b_in;
    Line x(n, 0.0);

    for (int k = 0; k < n; ++k) {
        int pivotRow = k;
        for (int i = k + 1; i < n; ++i)
            if (fabs(A[i][k]) > fabs(A[pivotRow][k]))
                pivotRow = i;
        swap(A[k], A[pivotRow]);
        swap(b[k], b[pivotRow]);

        for (int i = k + 1; i < n; ++i) {
            double factor = A[i][k] / A[k][k];
            for (int j = k; j < n; ++j)
                A[i][j] -= factor * A[k][j];
            b[i] -= factor * b[k];
        }
    }

    // Обратный ход
    for (int i = n - 1; i >= 0; --i) {
        double sum = b[i];
        for (int j = i + 1; j < n; ++j)
            sum -= A[i][j] * x[j];
        x[i] = sum / A[i][i];
    }
    return x;
}

Line zeidel(const Matrix& A, const Line& b) {
    int n = A.size();
    Line x(n, 0.0), prev(n, 0.0);
    bool printedLast = false;

    // Шапка таблицы
    cout << "\nМетод Зейделя: итерации каждые " << PRINT_STEP 
         << " шагов + финал\n";
    cout << setw(8) << "iter";
    for (int i = 0; i < n; ++i) cout << setw(15) << ("x"+to_string(i+1));
    cout << setw(15) << "eps\n";
    cout << string(8 + 15*n + 15, '-') << "\n";

    for (int iter = 1; iter <= MAX_ITER; ++iter) {
        // вычисляем новую итерацию
        for (int i = 0; i < n; ++i) {
            if (fabs(A[i][i]) < 1e-15) {
                cerr << "Ошибка: нулевой диагональный элемент!\n";
                exit(1);
            }
            double s = b[i];
            for (int j = 0; j < i; ++j) s -= A[i][j] * x[j];
            for (int j = i+1; j < n; ++j) s -= A[i][j] * prev[j];
            x[i] = s / A[i][i];
        }
        // считаем максимальную разницу
        double eps = 0;
        for (int i = 0; i < n; ++i)
            eps = max(eps, fabs(x[i] - prev[i]));

        // печатаем, если шаг совпадает или это финал
        if (iter % PRINT_STEP == 0 || eps < EPS) {
            cout << setw(8) << iter;
            for (double xi : x) cout << setw(15) << xi;
            cout << setw(15) << eps << "\n";
            if (eps < EPS) {
                printedLast = true;
                cout << "\nСошлось за " << iter
                     << " итераций (eps < " << EPS << ").\n";
                return x;
            }
        }

        prev = x;
    }

    if (!printedLast) {
        // если по какой-то причине не попали на финальный принт
        double eps = 0;
        for (int i = 0; i < n; ++i) eps = max(eps, fabs(x[i] - prev[i]));
        cout << setw(8) << MAX_ITER;
        for (double xi : x) cout << setw(15) << xi;
        cout << setw(15) << eps << "\n";
        cerr << "Не сошлось за " << MAX_ITER << " итераций!\n";
    }
    exit(1);
}

int main() {
    double M = 0.93, N = 0.07, P = -0.84;
    Matrix A = {
        {M, -0.04, 0.21, -18},
        {0.25, -1.23, N, -0.09},
        {-0.21, N, 0.8, -0.13},
        {0.15, -1.31, 0.06, P}
    };
    Line b = {-1.24, P, 2.56, M};

    cout << "Решение методом Гаусса:" << endl;
    for (auto i : gauss(A, b)) {
        cout << setw(6) << i << " "; // конкретизировать
    }
    cout << endl << "Решение методом Зейделя:" << endl;
    for (auto i : zeidel(A, b)) {
        cout << setw(6) << i << " ";
    }
    cout << endl;
}